pkg/streamingpromql/operators/selectors/extend_range_vector.go

```diff
@@ -9,89 +9,189 @@ import (
 	"github.com/prometheus/prometheus/promql"
 
 	"github.com/grafana/mimir/pkg/streamingpromql/types"
-	"github.com/grafana/mimir/pkg/util/limiter"
 )
 
+type undoAction int
+
+const (
+	none undoAction = iota
+	remove
+	replace
+)
+
+// AnchoredExtensionMetadata tracks the first and last points considered in a range extension.
+// It also tracks modifications made to the buffer, such as synthetic boundary points and head/tail trimming.
+// A function is provided to undo these changes on a given buffer.
+type AnchoredExtensionMetadata struct {
```

Comment:
See also my comment on `ConvertExtendedPointsToSmoothed`. If this is used by both anchored and smoothed, wouldn't renamed it to something more general will make clearer?